proteinworkshop/train.py

```diff
@@ -173,7 +175,7 @@ def train_model(
 
     if cfg.get("compile"):
         log.info("Compiling model!")
-        model = torch.compile(model)  # type: ignore
+        model = torch_geometric.compile(model, dynamic=True)
```

Comment:
Does this seem to work for any of the graph encoders?

Reply:
It works as in it runs but I experienced a slowdon of approx 2x. on SchNet. Only limited exploration though.

Reply:
Do you need to add the `fullgraph=True` flag to prevent any graph breaks from occurring? https://github.com/pyg-team/pytorch_geometric/blob/3e55a4c263f04ed6676618226f9a0aaf406d99b9/examples/compile/gin.py#L60

Reply:
For example, in Lightning 2.1, the recommended place to set this up for your `LightningModules` is:

```bash
@beartype
def setup(self, stage: str):
    """Lightning hook that is called at the beginning of fit (train + validate), validate,
    test, or predict.

    This is a good hook when you need to build models dynamically or adjust something about
    them. This hook is called on every process when using DDP.

    :param stage: Either `"fit"`, `"validate"`, `"test"`, or `"predict"`.
    """
    if self.hparams.compile and stage == "fit":
        if not torch_geometric.typing.WITH_PT21:
            raise Exception(
                "PyTorch Geometric's dynamic shape compilation requires PyTorch >= 2.1.0"
            )
        self.net = torch_geometric.compile(self.net, dynamic=True, fullgraph=True)
```

Reply:
In my experience, setting `fullgraph=True` can lead to downstream errors that are hard to debug, so I've had to leave it as `False` in the past.

Reply:
Ah, that's something to explore. I was getting graphbreak warnings (mostly torchmetrics-related).

Also, from the PyG docs (copied below) it seems we should pass batch size info to the pooling operators

```
It is generally a good practice to confirm that your written model does not contain any graph breaks. Importantly, there exists a few operations in PyG that will currently lead to graph breaks (but workaround exists), e.g.:

[global_mean_pool()](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.pool.global_mean_pool.html#torch_geometric.nn.pool.global_mean_pool) (and other pooling operators) perform device synchronization in case the batch size size is not passed, leading to a graph break.

[remove_self_loops()](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.remove_self_loops) and [add_remaining_self_loops()](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.add_remaining_self_loops) mask the given edge_index, leading to a device synchronization to compute its final output shape. As such, we recommend to augment your graph before inputting it into your GNN, e.g., via the [AddSelfLoops](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.transforms.AddSelfLoops.html#torch_geometric.transforms.AddSelfLoops) or [GCNNorm](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.transforms.GCNNorm.html#torch_geometric.transforms.GCNNorm) transformations, and setting add_self_loops=False/normalize=False when initializing layers such as GCNNorm.
```